Python/ceval.c

```diff
@@ -2476,14 +2476,42 @@ _PyEval_GetBuiltinId(_Py_Identifier *name)
 PyObject *
 PyEval_GetLocals(void)
 {
+    // We need to return a borrowed reference here, so some tricks are needed
     PyThreadState *tstate = _PyThreadState_GET();
      _PyInterpreterFrame *current_frame = _PyThreadState_GetFrame(tstate);
     if (current_frame == NULL) {
         _PyErr_SetString(tstate, PyExc_SystemError, "frame does not exist");
         return NULL;
     }
 
-    PyObject *locals = _PyEval_GetFrameLocals();
+    // Be aware that this returns a new reference
+    PyObject *locals = _PyFrame_GetLocals(current_frame);
+
+    if (locals == NULL) {
+        return NULL;
+    }
+
+    if (PyFrameLocalsProxy_Check(locals)) {
+        PyFrameObject *f = _PyFrame_GetFrameObject(current_frame);
+        PyObject *ret = PyDict_New();
```

Comment:
We don't actually want to make the snapshot in this case - we want to cache and return the proxy as described in PEP 667. It's the replacement for `PyFrame_LocalsToFast()` no longer doing anything.

Reply:
`PyEval_GetLocals()` returns a dict/mapping as `locals()`. Maybe you were thinking of `PyFrame_GetLocals()`?

Reply:
Turns out there is ambiguity on this point in PEP 667 (two references to caching a proxy instance, one to caching a snapshot): https://github.com/python/cpython/issues/118934#issuecomment-2143330127

Reply:
https://github.com/python/cpython/pull/119932 aligned the `PyEval_GetLocals` docs with the way it will work once this PR is completed and merged.